cmd/gateway/s3x/gateway-s3x-ledger-store.go

```diff
@@ -0,0 +1,178 @@
+package s3x
+
+import (
+	"context"
+	"sync"
+
+	pb "github.com/RTradeLtd/TxPB/v3/go"
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/namespace"
+)
+
+/* Design Notes
+---------------
+
+Internal functions should never claim or release locks.
+Any claiming or releasing of locks should be done in the public setter+getter functions.
+The reason for this is so that we can enable easy reuse of internal code.
+*/
+
+var (
+	dsPrefix    = datastore.NewKey("ledgerRoot")
+	dsBucketKey = datastore.NewKey("b") //bucket name to ipfsHash of LedgerBucketEntry
+	dsPartKey   = datastore.NewKey("p") //part ID to MultipartUpload
+)
+
+// ledgerStore is an internal bookkeeper that
+// maps buckets to ipfs cids and keeps a local cache of object names to hashes
+//
+// Bucket root hashes are saved in the provided data store.
+// Object hashes are saved in ipfs and cached in memory,
+// Object data is saved in ipfs.
+type ledgerStore struct {
+	ds  datastore.Batching
+	dag pb.NodeAPIClient //to be used as direct access to ipfs to optimize algorithm
+	l   *Ledger          //a cache of the values in datastore and ipfs
+
+	locker     bucketLocker //a locker to protect buckets from concurrent access (per bucket)
+	plocker    bucketLocker //a locker to protect MultipartUploads from concurrent access (per upload ID)
+	mapLocker  sync.Mutex   //a lock to protect the l.Buckets map from concurrent access
+	pmapLocker sync.Mutex   //a lock to protect the l.MultipartUploads map from concurrent access
+}
+
+func newLedgerStore(ds datastore.Batching, dag pb.NodeAPIClient) (*ledgerStore, error) {
+	ls := &ledgerStore{
+		ds:  namespace.Wrap(ds, dsPrefix),
+		dag: dag,
+		l: &Ledger{
+			Buckets:          make(map[string]*LedgerBucketEntry),
+			MultipartUploads: make(map[string]*MultipartUpload),
+		},
+	}
+	return ls, nil
+}
+
+func (ls *ledgerStore) getObjectHash(ctx context.Context, bucket, object string) (string, error) {
+	b, err := ls.getBucketLoaded(ctx, bucket)
+	if err != nil {
+		return "", err
+	}
+	objs := b.GetBucket().Objects
+	if objs == nil {
+		return "", ErrLedgerObjectDoesNotExist
+	}
+	h, ok := objs[object]
+	if !ok {
+		return "", ErrLedgerObjectDoesNotExist
+	}
+	return h, nil
+}
+
+func (ls *ledgerStore) object(ctx context.Context, bucket, object string) (*Object, error) {
+	h, err := ls.getObjectHash(ctx, bucket, object)
+	if err != nil {
+		return nil, err
+	}
+	return ipfsObject(ctx, ls.dag, h)
+}
+
+//ObjectInfo returns the ObjectInfo of the object.
+func (ls *ledgerStore) ObjectInfo(ctx context.Context, bucket, object string) (*ObjectInfo, error) {
```

Comment:
Why use a pointer? If we don't use a pointer the code can be cleaned up to be:
```Golang
defer ls.locker.read(bucket)()
return ls.object(ctx, bucket, object)
```

Reply:
Actually `ls.object` returns a point to `object`, not `ObjectInfo`. Otherwise, I want to avoid needlessly copying data.

Reply:
Good point